### ISSUE TITLE
Npm install fails when explicit npm version is already installed

### ISSUE THREAD

**Issue by michan85** [bitrise-steplib/steps-npm#23]
Im getting the following error running the task,it only appeared recently so it might be a bit rise issue.

i specify the version in the package.json then call the task with install command
```js
 "engines": {
    "npm": "6.4.1",
    "node": "10.15.3"
  }
```

```yaml
- npm@1.1.1:
        inputs:
        - command: install
        title: Npm Install
```

here is the error
```bash
+------------------------------------------------------------------------------+

| (3) Npm Install                                                              |
+------------------------------------------------------------------------------+
| id: npm                                                                      |
| version: 1.1.1                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2020-01-14T09:18:01Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
Config:
- Workdir: /Users/vagrant/git
- Command: install
- NpmVersion: 
- UseCache: false
Autodetecting npm version
Checking package.json for npm version
Ensuring npm version 6.4.1
$ npm "install" "-g" "npm@6.4.1"
Error setting npm version to 6.4.1: error running npm install: /usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
/usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
npm ERR! code EEXIST
npm ERR! syscall symlink
npm ERR! path ../../../lib/node_modules/npm/man/man5/npm-global.5
npm ERR! dest /usr/local/share/man/man5/npm-global.5
npm ERR! errno -17
npm ERR! EEXIST: file already exists, symlink '../../../lib/node_modules/npm/man/man5/npm-global.5' -> '/usr/local/share/man/man5/npm-global.5'
npm ERR! File exists: /usr/local/share/man/man5/npm-global.5
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vagrant/.npm/_logs/2020-01-14T09_18_16_714Z-debug.log
|                                                                              |
+---+---------------------------------------------------------------+----------+
| x | Npm Install (exit code: 1)                                    | 17.22 sec|
+---+---------------------------------------------------------------+----------+
| Issue tracker: https://github.com/bitrise-steplib/steps-npm/issues           |
| Source: https://github.com/bitrise-steplib/steps-npm                         |
+---+---------------------------------------------------------------+----------+
```

Im not sure but it seems like it could be avoided by just checking if its already installed then don't try install but its weird because this just started happening out of the blue so it might be bitrise related.

**Comment by anniewey** [comment]
Getting the same error since 13/01/2020 (running on `npm@1.1.1`). Was working fine previously and I didn't change any of the workflow/config.
```
Error setting npm version to 6.11.3: error running npm install: /usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
/usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
npm ERR! code EEXIST
npm ERR! syscall symlink
npm ERR! path ../../../lib/node_modules/npm/man/man5/npm-folders.5
npm ERR! dest /usr/local/share/man/man5/npm-folders.5
npm ERR! errno -17
npm ERR! EEXIST: file already exists, symlink '../../../lib/node_modules/npm/man/man5/npm-folders.5' -> '/usr/local/share/man/man5/npm-folders.5'
npm ERR! File exists: /usr/local/share/man/man5/npm-folders.5
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vagrant/.npm/_logs/2020-01-15T00_18_08_776Z-debug.log
```

**Comment by bitce** [comment]
Hey guys, thanks for reporting this! Could you please include build URLs so we can check out the full logs of these builds?

**Comment by anniewey** [comment]
Sure. Here's mine: https://app.bitrise.io/build/2ad970ced4987a79#?tab=log

**Comment by ahvth-bitrise** [comment]
@michan85 are you also using a MacOS stack?

**Comment by ahvth-bitrise** [comment]
@anniewey on macOS, this error can be avoided by using `--force` following the `-g` argument in your npm call.

**Comment by pkwiatkowski0** [comment]
Faced the same problem. I also tried @ahvth-bitrise workaround but not helped me. 
**logs**:
```
+------------------------------------------------------------------------------+

| (2) npm@1.1.1                                                                |
+------------------------------------------------------------------------------+
| id: npm                                                                      |
| version: 1.1.1                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: go                                                                  |
| time: 2020-01-20T13:05:11Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
Config:
- Workdir: ./
- Command: npm install -g --force
- NpmVersion: 6.13.4
- UseCache: false
Ensuring npm version 6.13.4
$ npm "install" "-g" "npm@6.13.4"
Error setting npm version to 6.13.4: error running npm install: /usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
/usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
npm ERR! code EEXIST
npm ERR! syscall symlink
npm ERR! path ../../../lib/node_modules/npm/man/man1/npm-access.1
npm ERR! dest /usr/local/share/man/man1/npm-access.1
npm ERR! errno -17
npm ERR! EEXIST: file already exists, symlink '../../../lib/node_modules/npm/man/man1/npm-access.1' -> '/usr/local/share/man/man1/npm-access.1'
npm ERR! File exists: /usr/local/share/man/man1/npm-access.1
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vagrant/.npm/_logs/2020-01-20T13_05_28_510Z-debug.log
|                                                                              |
+---+---------------------------------------------------------------+----------+
| x | npm@1.1.1 (exit code: 1)                                      | 19.39 sec|
+---+---------------------------------------------------------------+----------+
| Issue tracker: https://github.com/bitrise-steplib/steps-npm/issues           |
| Source: https://github.com/bitrise-steplib/steps-npm                         |
+---+---------------------------------------------------------------+----------+
```

**bitrise.yml**:
```
---
format_version: '8'
default_step_lib_source: https://github.com/bitrise-io/bitrise-steplib.git
project_type: cordova
trigger_map:
- push_branch: "*"
  workflow: primary
- pull_request_source_branch: "*"
  workflow: primary
workflows:
  primary:
    steps:
    - activate-ssh-key@4.0.5:
        run_if: '{{getenv "SSH_RSA_PRIVATE_KEY" | ne ""}}'
    - git-clone@4.0.18: {}
    - npm@1.1.1:
        inputs:
        - command: npm install -g --force
        - npm_version: 6.13.4
        - workdir: "$CORDOVA_WORK_DIR"
    - script@1.1.5:
        title: Do anything with Script step
        inputs:
        - content: |-
            #!/usr/bin/env bash
            # fail if any commands fails
            set -e
            # debug log
            set -x

            npm i
            npm i q
            npm i properties-parser
            npm i cordova-common
            npm i shelljs
            npm i nopt
            npm i compare-func
    - generate-cordova-build-configuration@0.9.6:
        inputs:
        - configuration: release
    - cordova-archive@2.1.0:
        inputs:
        - workdir: "$CORDOVA_WORK_DIR"
        - platform: "$CORDOVA_PLATFORM"
    - deploy-to-bitrise-io@1.9.5:
        inputs:
        - is_enable_public_page: 'false'
        - notify_user_groups: none
app:
  envs:
  - opts:
      is_expand: false
    CORDOVA_WORK_DIR: "./"
  - opts:
      is_expand: false
    CORDOVA_PLATFORM: android

```

**Comment by jtich** [comment]
Has anyone determined a solution besides attempting to use the `--force` flag?

**Comment by ahvth-bitrise** [comment]
@jtich @pkwiatkowski0 you could also try manually removing the packages that are causing the conflict beforehand. Our tooling team is currently looking into this, though, so hopefully we can roll out a fix soon. Please let me know if you have any issues with that approach.

**Comment by ahvth-bitrise** [comment]
Hi everyone! 👋 It's likely that this is a side effect of preinstalling node using brew. Can you please try issuing the following before your npm step within a script step?

```
brew link --overwrite node
```

Let us know if that helps! 😄

**Comment by listenrightmeow** [comment]
@ahvth-bitrise I just implemented the suggested fix, NPM still failed.

`Linking /usr/local/Cellar/node/11.12.0... 686 symlinks created`

```
Config:
- Workdir: /Users/vagrant/git
- Command: install
- NpmVersion: 6.10.1
- UseCache: false
Ensuring npm version 6.10.1
$ npm "install" "-g" "npm@6.10.1"
Error setting npm version to 6.10.1: error running npm install: /usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
/usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
npm ERR! code EEXIST
npm ERR! syscall symlink
npm ERR! path ../../../lib/node_modules/npm/man/man5/npm-folders.5
npm ERR! dest /usr/local/share/man/man5/npm-folders.5
npm ERR! errno -17
npm ERR! EEXIST: file already exists, symlink '../../../lib/node_modules/npm/man/man5/npm-folders.5' -> '/usr/local/share/man/man5/npm-folders.5'
npm ERR! File exists: /usr/local/share/man/man5/npm-folders.5
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vagrant/.npm/_logs/2020-01-22T14_19_30_972Z-debug.log
```

**Comment by listenrightmeow** [comment]
As part of debugging, I introduced two NPM commands prior to install (prune & uninstall). I noticed that our pipeline was running mis-matched NPM versions for the debug steps stated above.

The NPM failure on our end was a 6.10.x version issue. I suggest removing the hard-coded $npm_version if it's being utilized to bypass this issue.

**Comment by non-binary** [comment]
Hi @michan85 @anniewey @pkwiatkowski0 ,

Thank you for your patience. I have reached out to our infrastructure team internally to see what they suggest. I will follow-up here as soon as I have heard back from them.

Thanks,
Luna

**Comment by grifotv** [comment]
I'm having the exact same error. I can confirm it started when I changed my workflow the stack from `Xcode 10.3, on macOS 10.14 (Mojave)` to `Xcode 11.3x, on macOS 10.14.6 (Mojave)`.

**Comment by grifotv** [comment]
As I need Xcode 11, I ended up using the stack `Xcode 11.0x, on macOS 10.14.5 (Mojave)` and it works fine now.  🎉

**Comment by non-binary** [comment]
Hi @grifotv,

Thank you for letting us know that changing the stack worked for you.

To help us resolve the issue on the latest stacks, can you please provide me with a fresh build log of your project running on the broken stack and one on the working stack? If your project is private, feel free to reach out to letsconnect [at] bitrise.io and reference this GH issue and me (my name is Luna).

Cheers and looking forward to receiving those logs.
Luna

**Comment by grifotv** [comment]
@non-binary how can I get the "fresh build log"? Can you point me to the right direction? I'm not sure how to do it. Thanks

**Comment by non-binary** [comment]
@grifotv I mean to run a new/recent build with the broken stack to make sure the build still fails. As well as a new/recent build with the working stack to make sure that build still succeeds. Thanks!

**Comment by lpusok** [comment]
Hello @michan85, @anniewey, @jtich, @listenrightmeow, @grifotv

So far the issue seems specific to the brew releases of node.js LTS 10 and 12 releases bundled with npm 6.13.4.
The following script was tested as a workaround:
```
brew unlink node@10
brew install node
brew link node --overwrite --force
```

We are looking into updating node or installing it from an other sources.

**Comment by lpusok** [comment]
Hello @michan85, @anniewey, @jtich, @listenrightmeow, @grifotv
Released a new version of the npm step (1.1.2: https://github.com/bitrise-steplib/steps-npm/releases/tag/1.1.2), which includes a fix.
Please let us know if you still experience this problem.
